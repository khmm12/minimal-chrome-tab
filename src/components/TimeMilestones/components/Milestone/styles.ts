import { css } from 'styled-system/css'
import * as p from 'styled-system/patterns'

export const container = p.vstack.raw({
  alignItems: 'center',
  gap: 0,
})

export const value = css.raw({
  fontSize: { base: '4rem', smDown: '3.375rem' },
})

export const description = css.raw({
  margin: '0.625rem 0 0',
  fontSize: { base: '1.5rem', smDown: '1rem' },
  textTransform: 'lowercase',
})
