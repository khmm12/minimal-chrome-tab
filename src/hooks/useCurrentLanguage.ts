import type { Accessor } from 'solid-js'
import getCurrentLocale from '@/utils/get-current-locale'

export default function useCurrentLanguage(): Accessor<string> {
  const language = getCurrentLocale()
  return () => language
}
