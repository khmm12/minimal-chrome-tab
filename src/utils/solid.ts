export type SignalValue<T> = () => T

export function unwrap<T>(mayBeSingnal: SignalValue<T> | Exclude<T, Function>): T {
  // @ts-expect-error
  return typeof mayBeSingnal === 'function' ? mayBeSingnal() : mayBeSingnal
}
