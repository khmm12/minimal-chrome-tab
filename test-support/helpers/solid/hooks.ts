import { createRoot } from 'solid-js'

type Disposal = () => void

const disposals = new Set<Disposal>()

afterEach(() => cleanup())

export function renderHook<T>(fn: () => T): T {
  return withRoot(fn)[0]
}

export function withRoot<T>(fn: () => T): [value: T, dispose: () => void] {
  let disposal: () => void

  const value = createRoot((dispose) => {
    disposal = dispose
    disposals.add(dispose)
    return fn()
  })

  const dispose = (): void => {
    disposal()
    disposals.delete(disposal)
  }

  return [value, dispose]
}

export function cleanup(): void {
  for (const dispose of disposals) dispose()
  disposals.clear()
}
