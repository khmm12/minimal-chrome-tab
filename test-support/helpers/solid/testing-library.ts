import type { JSX } from 'solid-js'
import { hydrate as solidHydrate, render as solidRender } from 'solid-js/web'
import { getQueriesForElement, prettyDOM, queries as allQueries } from '@testing-library/dom'
import type { Queries, BoundFunction, prettyFormat } from '@testing-library/dom'

interface Ref {
  container: HTMLElement
  unmount: () => void
  dispose: () => void
}

type Ui = () => JSX.Element

interface Options<Q extends Queries = typeof allQueries> {
  container?: HTMLElement
  baseElement?: HTMLElement
  queries?: Q
  hydrate?: boolean
}

type DebugFn = (
  baseElement?: HTMLElement | HTMLElement[],
  maxLength?: number,
  options?: prettyFormat.OptionsReceived
) => void

type Result<Q extends Queries = typeof allQueries> = {
  container: HTMLElement
  baseElement: HTMLElement
  debug: DebugFn
  unmount: () => void
} & { [P in keyof Q]: BoundFunction<Q[P]> }

afterEach(() => cleanup())

const mountedContainers = new Set<Ref>()

function render<Q extends Queries = typeof allQueries>(ui: Ui, options: Options<Q>): Result<Q>
function render(ui: Ui, options?: Omit<Options, 'queries'>): Result

function render(ui: Ui, options: Options = {}): unknown {
  let { container, baseElement = container ?? document.body, queries, hydrate = false } = options

  container ??= baseElement.appendChild(document.createElement('div'))

  const unmount = (): void => {
    if (container?.parentElement === document.body) container.remove()
  }

  const dispose = hydrate ? solidHydrate(ui, container) : solidRender(ui, container)

  mountedContainers.add({ container, unmount, dispose })

  const debug: DebugFn = (el = baseElement, maxLength, options) =>
    Array.isArray(el)
      ? el.forEach((e) => console.log(prettyDOM(e, maxLength, options)))
      : console.log(prettyDOM(el, maxLength, options))

  return {
    container,
    baseElement,
    debug,
    unmount: dispose,
    ...getQueriesForElement(baseElement, queries),
  }
}

function cleanup(): void {
  mountedContainers.forEach(cleanupAtContainer)
  mountedContainers.clear()
}

function cleanupAtContainer(ref: Ref): void {
  ref.dispose()
  ref.unmount()
  mountedContainers.delete(ref)
}

export { render, cleanup }
