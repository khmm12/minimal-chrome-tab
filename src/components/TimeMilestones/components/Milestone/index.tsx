import type { JSX } from 'solid-js'
import { css } from 'styled-system/css'
import createIntlFormatter from '@/hooks/createIntlFormatter'
import { round } from '@/utils/rounds'
import Bar from './components/Bar'
import * as s from './styles'

interface MilestoneProps {
  value: number
  description: string
}

const PercentFormatOptions = {
  style: 'percent',
  minimumFractionDigits: 0,
  maximumFractionDigits: 0,
} satisfies Intl.NumberFormatOptions

export default function Milestone(props: MilestoneProps): JSX.Element {
  const format = createIntlFormatter()

  const formatPercent = (value: number): string => format.number(value, PercentFormatOptions)
  const formatValue = (value: number): string => formatPercent(round(value, 2)).replaceAll(/\s/g, '')

  return (
    <div role="group" aria-label={props.description} class={css(s.container)}>
      <span class={css(s.value)}>{formatValue(props.value)}</span>
      <Bar progress={props.value} />
      <span class={css(s.description)}>
        <span aria-hidden="true">...</span>
        {props.description}
      </span>
    </div>
  )
}
