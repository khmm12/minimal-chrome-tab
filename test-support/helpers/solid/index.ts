import { render, cleanup as cleanupElements } from './testing-library'
import { renderHook, withRoot, cleanup as cleanupHooks } from './hooks'

export { renderHook, withRoot, render }

export function cleanup(): void {
  cleanupHooks()
  cleanupElements()
}

export * from '@testing-library/dom'
