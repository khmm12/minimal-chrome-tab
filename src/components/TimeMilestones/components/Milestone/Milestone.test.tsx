import { render, screen } from '@test/helpers/solid'
import useCurrentLanguage from '@/hooks/useCurrentLanguage'
import Milestone, { MilestoneVariant } from '.'

vi.mock('@/hooks/useCurrentLanguage')

beforeEach(() => {
  vi.mocked(useCurrentLanguage).mockReturnValue(() => 'en-US')
})

afterEach(() => {
  vi.resetAllMocks()
})

describe('Milestone', () => {
  it('renders correctly', () => {
    render(() => <Milestone variant={MilestoneVariant.BarsCompact} value={0.25} description="test milestone" />)

    const r = screen.getByLabelText('test milestone')

    expect(r).toBeInTheDocument()
    expect(r).toMatchSnapshot()
  })
})
