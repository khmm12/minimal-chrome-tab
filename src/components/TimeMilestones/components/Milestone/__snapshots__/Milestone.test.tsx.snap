// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Milestone > renders correctly 1`] = `
<div
  aria-label="test milestone"
  class="d_flex ai_center gap_0 flex-d_column"
  role="group"
>
  <span
    class="fs_4rem smDown:fs_3.375rem"
  >
    25%
  </span>
  <svg
    aria-hidden="true"
    height="1rem"
    preserveAspectRatio="xMinYMin"
    role="figure"
    viewBox="0 0 25 5"
  >
    <path
      d="M0.6 0 L0.6 5 Z"
      stroke="currentColor"
      stroke-width="1.2"
    />
    <path
      d="M3.2444444444 0 L3.2444444444 5 Z"
      stroke="currentColor"
      stroke-width="1.2"
    />
    <path
      d="M5.8888888889 2.375 L5.8888888889 5 Z"
      stroke="currentColor"
      stroke-width="1.2"
    />
    <path
      d="M8.5333333333 4.75 L8.5333333333 5 Z"
      stroke="currentColor"
      stroke-width="1.2"
    />
    <path
      d="M11.1777777778 4.75 L11.1777777778 5 Z"
      stroke="currentColor"
      stroke-width="1.2"
    />
    <path
      d="M13.8222222222 4.75 L13.8222222222 5 Z"
      stroke="currentColor"
      stroke-width="1.2"
    />
    <path
      d="M16.4666666667 4.75 L16.4666666667 5 Z"
      stroke="currentColor"
      stroke-width="1.2"
    />
    <path
      d="M19.1111111111 4.75 L19.1111111111 5 Z"
      stroke="currentColor"
      stroke-width="1.2"
    />
    <path
      d="M21.7555555556 4.75 L21.7555555556 5 Z"
      stroke="currentColor"
      stroke-width="1.2"
    />
    <path
      d="M24.4 4.75 L24.4 5 Z"
      stroke="currentColor"
      stroke-width="1.2"
    />
  </svg>
  <span
    class="m_0.625rem_0_0 fs_1.5rem smDown:fs_1rem tt_lowercase"
  >
    <span
      aria-hidden="true"
    >
      ...
    </span>
    test milestone
  </span>
</div>
`;
